sqlcapture/capture.go

```diff
@@ -250,35 +250,47 @@ func (c *Capture) updateState(ctx context.Context) error {
 	// initialize new state entries for these streams, and while we're at
 	// it this is a good time to sanity-check the primary key configuration.
 	for streamID, binding := range c.Bindings {
-		// If the `PrimaryKey` property is specified in the catalog then use that,
-		// otherwise use the "native" primary key of this table in the database.
-		// Print a warning if the two are not the same.
-		var primaryKey []string
 		var discoveryInfo = c.discovery[streamID]
 		if discoveryInfo == nil {
 			return fmt.Errorf("table %q is a configured binding of this capture, but doesn't exist or isn't visible with current permissions", streamID)
 		}
-		if len(discoveryInfo.PrimaryKey) > 0 {
-			primaryKey = discoveryInfo.PrimaryKey
-			if len(primaryKey) != 0 {
-				logrus.WithFields(logrus.Fields{
-					"table": streamID,
-					"key":   primaryKey,
-				}).Debug("queried primary key")
-			}
-		}
-		if len(binding.Resource.PrimaryKey) != 0 {
-			if strings.Join(primaryKey, ",") != strings.Join(binding.Resource.PrimaryKey, ",") {
-				logrus.WithFields(logrus.Fields{
-					"stream":      streamID,
-					"catalogKey":  binding.Resource.PrimaryKey,
-					"databaseKey": primaryKey,
-				}).Warn("primary key in catalog differs from database table")
-			}
+
+		// Select the primary key from the first available source. In order of priority:
+		//   - If the resource config specifies a primary key override then we'll use that.
+		//   - Normally we'll use the primary key from the collection spec.
+		//   - And as a fallback which *should* no longer be reachable, we'll use the
+		//     primary key from the startup discovery run.
+		logrus.WithFields(logrus.Fields{
+			"stream":     streamID,
+			"resource":   binding.Resource.PrimaryKey,
+			"collection": binding.CollectionKey,
+			"discovery":  discoveryInfo.PrimaryKey,
+		}).Debug("selecting primary key")
+		var primaryKey []string
+		if len(binding.Resource.PrimaryKey) > 0 {
 			primaryKey = binding.Resource.PrimaryKey
+			logrus.WithFields(logrus.Fields{"stream": streamID, "key": primaryKey}).Debug("using resource primary key")
+		} else if len(binding.CollectionKey) > 0 {
+			// TODO(wgd): This logic isn't correct for all possible column names, but
```

Comment:
Sorry, that comment is rather vague. The edge cases I'm concerned about are described in https://github.com/estuary/connectors/issues/620 -- basically we need to decode the `~0` and `~1` escapes into `~` and `/` respectively. So this logic as currently written is incompatible with column names containing slashes, essentially.

**However**, our discovery logic is similarly broken because it just prefixes a slash onto the unescaped column name, and AFAIK what will happen then is:
1. We observe a column named `"foo/bar"` in the database and emit a discovery result with key `/foo/bar`.
2. Flow sees a name like `/foo/bar` and interprets it as a nested property, and then immediately gets upset because property `/foo/bar` is not present and marked required in the schema.
3. Even if we somehow got past point 2 the capture would fail the first time it tried to emit a document because that property would be missing.

So I don't believe that merging this PR as-is breaks things any worse than they already are in any situation, but I intend to come back and make slashes and tildes in primary keys actually get escaped and then unescaped correctly in the near future.